### PR TITLE
verifies shred slot and parent in fetch stage

### PR DESCRIPTION
**Pull request by behzadnouri**
#### Problem
Shred slot and parent are not verified until window-service where
resources are already wasted to sig-verify and deserialize shreds.

#### Summary of Changes
This commit moves above verification to earlier in the pipeline in fetch
stage.

### CODE REVIEW

File: ledger/src/shred.rs [hunk 1]
```diff
@@ -573,6 +573,15 @@ pub mod layout {
             .ok()
     }
 
+    // The caller should verify first that the shred is data and not code!
+    pub(super) fn get_parent_offset(shred: &[u8]) -> Option<u16> {
+        const OFFSET_OF_SHRED_PARENT: usize = SIZE_OF_COMMON_SHRED_HEADER;
+        debug_assert_eq!(get_shred_type(shred).unwrap(), ShredType::Data);
+        <[u8; 2]>::try_from(shred.get(OFFSET_OF_SHRED_PARENT..)?.get(..2)?)
```
Comment (jbiseda) [inline ledger/src/shred.rs:580]: nit: `OFFSET_OF_SHRED_PARENT..OFFSET_OF_SHRED_PARENT + 2`

**Review by jbiseda — APPROVED** [review]
looks good

File: ledger/src/shred.rs [hunk 2]
```diff
@@ -653,48 +662,95 @@ impl TryFrom<u8> for ShredVariant {
...
+                Some(parent_offset) => parent_offset,
+                None => {
+                    stats.bad_parent_offset += 1;
+                    return true;
+                }
+            };
+            if parent_offset == 0 && slot != 0 {
+                stats.bad_parent_offset += 1;
+                return true;
+            }
```
Comment (steviez) [inline ledger/src/shred.rs:739]: We have some slightly duplicated checking happening on data shreds:
1) We error if `slot <= root` on line 694 (all shreds)
2) We later error `if parent_offset == 0 && slot != 0` on line 736 (data shreds only)
3) The comparisons from 1. and 2., plus some extras are checked in `blockstore::verify_shred_slots()` on line 747 (data shreds only)
```
    if slot == 0 && parent == 0 && root == 0 {
        return true; // valid write to slot zero.
    }
    // Ignore shreds that chain to slots before the root,
    // or have invalid parent >= slot.
    root <= parent && parent < slot
```
So, I think we could:
- Drop the check in 2.
- Move the check to ensure `slot > root` from 694 to only happen on coding shreds, as `verify_shred_slots()` will confirm `slot > root` for data shreds

I can throw up a small PR for this later

**Review by steviez — COMMENTED** [review]
Sorry about the late review - things look good to me, just one micro optimization I think we can make

File: ledger/src/shred.rs (hunk 2 above)
Comment (behzadnouri) [reply ledger/src/shred.rs:739]: @steviez yeah, that sounds right, thanks

**Pull request merged**